### DISCUSSION TITLE
Does hubs process video to cater to low bandwidth users?

### DISCUSSION THREAD

**Discussion by SeanBannister** [mozilla/hubs#2696] [General Questions & Discussion] [answered]
When videos are uploaded to Hubs/Spoke are they processed in anyway such as creating multiple resolutions or streaming them back to the user using adaptive bitrate? What about when a user live streams their desktop using hubs? I'm creating some rooms for users on low bandwidth devices and I'm considering what resolution I should upload.

I did see #2483 [Reticulum doesn't care about client network conditions for HLS Adaptive Bitrate Streaming](https://github.com/mozilla/hubs/issues/2483).

**Comment by emclaren** [comment]
To my knowledge, the video streams themselves are not adaptive. There are some performance optimizations on low bandwidth devices. For example, you can go into the user preferences in a hubs room on an individual device and reduce the Hubs room resolution (this should happen automatically if your frame rate is low for a sustained period of time). Some types of lighting are also limited on mobile to help improve performance. 

Improving performance on mobile devices is something we are interested in exploring more in the next little while ( https://github.com/mozilla/hubs/issues/2638 ), I don't know if videos will be included in this but it is possible. 

I'm not sure about what resolution would be best for your use case (there is obviously a trade off between resolution & performance), it depends a bit on how many people you are expecting to be in the room at the same time. If there will only be a few people at once, a larger video should be fine. But if you are expecting closer to 25 at once I would definitely recommend keeping the file size small. It might be worthwhile doing a test run in advance if possible (note: we have weekly meetups and are open to touring community member's spaces to give feedback on scene performance, feel free to reach out with the request in the #show-and-tell channel on our Discord Server).

**Comment by netpro2k** [comment, marked as answer]
We don't do any sort of video processing on the backend, but If you have a direct link to a HLS or DASH playlist with adaptive bitrate settings they should work correctly in Hubs, though as you pointed out there may be some bugs. If you have an example of a stream you expect to be using adaptive bitrate but it is not, providing a link would be helpful. 

Under the hood we are using hls.js and dash.js libraries to do the playback, largely with default settings.
